pkg/quota/controller/clusterquotamapping/clusterquotamapping.go

```diff
@@ -0,0 +1,585 @@
+
+		clusterQuotaMapper: &clusterQuotaMapper{
+			requiredQuotaToSelector:    map[string]*unversioned.LabelSelector{},
+			requiredNamespaceToLabels:  map[string]map[string]string{},
+			completedQuotaToSelector:   map[string]*unversioned.LabelSelector{},
+			completedNamespaceToLabels: map[string]map[string]string{},
+
+			quotaToNamespaces: map[string]sets.String{},
+			namespaceToQuota:  map[string]sets.String{},
+		},
+	}
+
+	namespaceInformer.Informer().AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addNamespace,
+		UpdateFunc: c.updateNamespace,
+		DeleteFunc: c.deleteNamespace,
+	})
+	c.namespaceLister = namespaceInformer.Lister()
+	c.namespacesSynced = namespaceInformer.Informer().HasSynced
+
+	quotaInformer.Informer().AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addQuota,
+		UpdateFunc: c.updateQuota,
+		DeleteFunc: c.deleteQuota,
+	})
+	c.quotaLister = quotaInformer.Lister()
+	c.quotasSynced = quotaInformer.Informer().HasSynced
+
+	return c
+}
+
+type ClusterQuotaMappingController struct {
+	namespaceQueue   workqueue.RateLimitingInterface
+	namespaceLister  *ocache.IndexerToNamespaceLister
+	namespacesSynced func() bool
+
+	quotaQueue   workqueue.RateLimitingInterface
+	quotaLister  *ocache.IndexerToClusterResourceQuotaLister
+	quotasSynced func() bool
+
+	clusterQuotaMapper *clusterQuotaMapper
+}
+
+type clusterQuotaMapper struct {
+	lock sync.RWMutex
+
+	requiredQuotaToSelector    map[string]*unversioned.LabelSelector
+	requiredNamespaceToLabels  map[string]map[string]string
+	completedQuotaToSelector   map[string]*unversioned.LabelSelector
+	completedNamespaceToLabels map[string]map[string]string
+
+	quotaToNamespaces map[string]sets.String
+	namespaceToQuota  map[string]sets.String
+}
+
+func (m *clusterQuotaMapper) GetClusterQuotasFor(namespaceName string) ([]string, map[string]string) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	quotas, ok := m.namespaceToQuota[namespaceName]
+	if !ok {
+		return []string{}, m.completedNamespaceToLabels[namespaceName]
+	}
+	return quotas.List(), m.completedNamespaceToLabels[namespaceName]
+}
+
+func (m *clusterQuotaMapper) GetNamespacesFor(quotaName string) ([]string, *unversioned.LabelSelector) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	namespaces, ok := m.quotaToNamespaces[quotaName]
+	if !ok {
+		return []string{}, m.completedQuotaToSelector[quotaName]
+	}
+	return namespaces.List(), m.completedQuotaToSelector[quotaName]
+}
+
+// requireQuota changes the updates the selector requirements for the given quota, returns true
+// if a modification was made
+func (m *clusterQuotaMapper) requireQuota(quota *quotaapi.ClusterResourceQuota) bool {
+	m.lock.RLock()
+	selector, exists := m.requiredQuotaToSelector[quota.Name]
+	m.lock.RUnlock()
+
+	if selectorMatches(selector, exists, quota) {
+		return false
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	selector, exists = m.requiredQuotaToSelector[quota.Name]
```

Comment:
> What's the difference between this and the check above?

Two different lock acquisitions.  Read locks are shareable, so this code is 
 1. acquire shareable lock to read data
 1. release lock
 1. check data.  if you don't match, return early
 1. we might do work, so acquire a non-shareable lock
 1. read and check the data again now that we have a write lock